src/rpc/blockchain.cpp

```diff
@@ -1203,6 +1202,89 @@ UniValue getchaintips(const UniValue& params, bool fHelp)
+            "    \"hash\": \"xxxx\",         (string) block hash of the tip\n"
+            "    \"height\": xxxx,         (numeric) height of the chain tip\n"
+            "    \"sprout\": {\n"
+            "       \"finalCommitmentRoot\": \"hex\" (string)\n"
+            "       \"finalCommitmentState\": \"hex\" (string)\n"
+            "       \"skipHash\": \"hash\" (string)\n"
+            "     },\n"
+            "    \"sapling\": {\n"
+            "       \"finalCommitmentRoot\": \"hex\" (string)\n"
+            "       \"finalCommitmentState\": \"hex\" (string)\n"
+            "       \"skipHash\": \"hash\" (string)\n"
+            "     }\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_gettreestate", "\"00000000febc373a1da2bd9f887b105ad79ddc26ac26c2b28652d64e5207c5b5\"")
+            + HelpExampleRpc("z_gettreestate", "\"00000000febc373a1da2bd9f887b105ad79ddc26ac26c2b28652d64e5207c5b5\"")
+            + HelpExampleCli("z_gettreestate", "12800")
+            + HelpExampleRpc("z_gettreestate", "12800")
+        );
+
+    LOCK(cs_main);
+
+    std::string strHash = params[0].get_str();
+
+    // If height is supplied, find the hash
+    if (strHash.size() < (2 * sizeof(uint256))) {
+        strHash = chainActive[parseHeightArg(strHash)]->GetBlockHash().GetHex();
+    }
+
+    uint256 hash(uint256S(strHash));
+
+    if (mapBlockIndex.count(hash) == 0)
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
+    CBlockIndex* pblockindex = mapBlockIndex[hash];
+
+    UniValue res(UniValue::VOBJ);
+    res.pushKV("hash", pblockindex->GetBlockHash().GetHex());
+    res.pushKV("height", pblockindex->nHeight);
+    res.pushKV("time", int64_t(pblockindex->nTime));
+
+    // sprout
+    {
+        UniValue sprout_result(UniValue::VOBJ);
+        sprout_result.pushKV("finalCommitmentRoot", pblockindex->hashFinalSproutRoot.GetHex());
+        SproutMerkleTree tree;
+        if (!pcoinsTip->GetSproutAnchorAt(pcoinsTip->GetBestAnchor(SPROUT), tree)) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Cannot get sprout commitment tree");
+        }
+        CDataStream s(SER_NETWORK, PROTOCOL_VERSION);
+        s << tree;
+        sprout_result.pushKV("finalCommitmentState", HexStr(s.begin(), s.end()));
+        sprout_result.pushKV("skipHash", "");
+        res.pushKV("sprout", sprout_result);
+    }
+
+    // sapling
+    {
+        UniValue sapling_result(UniValue::VOBJ);
+        sapling_result.pushKV("finalCommitmentRoot", pblockindex->hashFinalSaplingRoot.GetHex());
+        SaplingMerkleTree tree;
+        if (!pcoinsTip->GetSaplingAnchorAt(pblockindex->hashFinalSaplingRoot, tree)) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Cannot get sapling commitment tree");
+        }
+        CDataStream s(SER_NETWORK, PROTOCOL_VERSION);
+        s << tree;
+        sapling_result.pushKV("finalCommitmentState", HexStr(s.begin(), s.end()));
+        sapling_result.pushKV("skipHash", "");
```

Comment:
Same changes here (but we can short-circuit the skiplist when we reach Sapling activation height):
```suggestion
        if (pcoinsTip->GetSaplingAnchorAt(pblockindex->hashFinalSaplingRoot, tree)) {
            CDataStream s(SER_NETWORK, PROTOCOL_VERSION);
            s << tree;
            sapling_result.pushKV("finalCommitmentState", HexStr(s.begin(), s.end()));
        }
        auto pindexSkip = pblockindex->pprev;
        while (pindexSkip && consensusParams.NetworkUpgradeActive(pindexSkip->nHeight, Consensus::UPGRADE_SAPLING)) {
            if (pcoinsTip->GetSaplingAnchorAt(pindexSkip->hashFinalSaplingRoot, tree)) {
                sapling_result.pushKV("skipHash", pindexSkip->GetBlockHash().GetHex());
                break;
            } else {
                pindexSkip = pblockindex->pprev;
            }
        }
```